src/derive/field.rs

```diff
@@ -429,31 +430,69 @@ macro_rules! field_arithmetic {
             }
 
             /// Multiplies `rhs` by `self`, returning the result.
-            #[inline]
-            pub const fn mul(&self, rhs: &Self) -> $field {
-                // Schoolbook multiplication
+            #[inline(always)]
+            #[unroll::unroll_for_loops]
+            #[allow(unused_assignments)]
+            pub const fn mul(&self, rhs: &Self) -> Self {
+                // Fast Coarsely Integrated Operand Scanning (CIOS) as described
+                // in Algorithm 2 of EdMSM: https://eprint.iacr.org/2022/1400.pdf
+                //
+                // Cannot use the fast version (algorithm 2) if
+                // modulus_high_word >= (WORD_SIZE - 1) / 2 - 1 = (2^64 - 1)/2 - 1
+
+                if $modulus.0[3] < (u64::MAX / 2) {
+                    const N: usize = 4;
+                    let mut t: [u64; N] = [0u64; N];
+                    let mut c_2: u64;
+                    for i in 0..4 {
+                        let mut c: u64 = 0u64;
+                        for j in 0..4 {
+                            (t[j], c) = mac(t[j], self.0[j], rhs.0[i], c);
+                        }
+                        c_2 = c;
+
+                        let m = t[0].wrapping_mul(INV);
+                        (_, c) = macx(t[0], m, $modulus.0[0]);
+
+                        for j in 1..4 {
+                            (t[j - 1], c) = mac(t[j], m, $modulus.0[j], c);
+                        }
+                        (t[N - 1], _) = adc(c_2, c, 0);
+                    }
+
+                    if bigint_geq(&t, &$modulus.0) {
```

Comment:
It seems the `bigint_geq` procedure and its usage may be suboptimal. First of all, you can notice that $t ≥ m ⇔ !(t < m)$. You can implement the "less-than" as subtraction $t - m$. If it overflows (borrow is 1) then $t < m$.
Then you can notice that you are actually computing the $t - m$ just after. The common pattern is to compute the subtraction speculatively and use its result in case it hasn't overflowed.

```
(tmp, borrow) = t - m
t = borrow ? t : tmp

Reply:
Nice catch!

I believe strategy would only save instructions in the case that `bigint_geq == true`. It would cost some instructions in the case that `bigint_geq == false` as `sbb` gets broken out into a few instructions on ARM where as the 4 u64 LTs should be a single instruction each.